storage_scrubber/src/checks.rs

```diff
@@ -527,3 +531,136 @@ async fn list_timeline_blobs_impl(
+
+    let mut stream = std::pin::pin!(stream_listing(remote_client, &tenant_root_target));
+    'outer: while let Some(obj) = stream.next().await {
+        let (key, Some(obj)) = obj? else {
+            panic!("ListingObject not specified");
+        };
+
+        let blob_name = key.get_path().as_str().strip_prefix(prefix_str);
+        'err: {
+            // TODO a let chain would be nicer here.
+            let Some(name) = blob_name else {
+                break 'err;
+            };
+            if !name.starts_with("tenant-manifest") {
+                break 'err;
+            }
+            let Ok(path) = RemotePath::from_string(name) else {
+                break 'err;
+            };
+            let Some(generation) = parse_remote_tenant_manifest_path(path) else {
+                break 'err;
+            };
+            tracing::debug!("tenant manifest {key}");
+            manifests.push((generation, obj));
+            continue 'outer;
+        }
+        tracing::info!("Listed an unknown key: {key}");
+        unknown_keys.push(obj);
+    }
+
+    if manifests.is_empty() {
+        tracing::debug!("No manifest for timeline.");
+
+        return Ok(ListTenantManifestResult {
+            errors,
+            manifest_info: RemoteTenantManifestInfo {
+                latest_generation: None,
+                manifests,
+                unknown_keys,
+            },
+        });
+    }
+
+    // Find the manifest with the highest generation
+    let (latest_generation, latest_listing_object) = manifests
+        .iter()
+        .max_by_key(|i| i.0)
+        .map(|(g, obj)| (*g, obj.clone()))
+        .unwrap();
+
+    let manifest_bytes =
+        match download_object_with_retries(remote_client, &latest_listing_object.key).await {
+            Ok(bytes) => bytes,
+            Err(e) => {
+                // It is possible that the tenant gets deleted in-between we list the objects
+                // and we download the manifest file.
+                errors.push(format!("failed to download tenant-manifest.json: {e}"));
+                return Ok(ListTenantManifestResult {
+                    errors,
+                    manifest_info: RemoteTenantManifestInfo {
+                        latest_generation: Some(latest_generation),
+                        manifests,
+                        unknown_keys,
+                    },
+                });
+            }
+        };
+
+    match TenantManifest::from_json_bytes(&manifest_bytes) {
+        Ok(_manifest) => {
+            return Ok(ListTenantManifestResult {
+                errors,
+                manifest_info: RemoteTenantManifestInfo {
+                    latest_generation: Some(latest_generation),
+                    manifests,
+                    unknown_keys,
+                },
+            });
+        }
+        Err(parse_error) => errors.push(format!(
+            "tenant-manifest.json body parsing error: {parse_error}"
+        )),
```

Comment:
there could theoretically be multiple errors per key, so a `Vec<(RemotePath, String)>` is better. In general, we shouldn't spend too much time on this because it's "just" logged later on anyways, where we include the tenant id.

Reply:
I mean, maybe it's useful for later refactors to put it into some kind of report that we put into the storcon db.